### DISCUSSION TITLE
Mise is aware of, but doesn't apply env variable expansions from multiple files

### DISCUSSION THREAD

**Discussion by jmoppel** [jdx/mise#3897] [Troubleshooting and bug reports]
I have an issue with variable expansion in .env files. I don't think it is a dotenvy problem. Essentially, two of of my environment variables are blank that use env variable expansion. However, when I issue mise envs, it shows the correctly expanded values.

I have two env files `.postgres` and `.local`.
```toml
[env]
_.file = [
    ".envs/.dev/.postgres",
    ".envs/.dev/.local",
]
```

```
# .postgres
PGHOST=postgres
```

```
# .local
PGTEST="${PGHOST}"
PGTESTING=$PGHOST
```

The PGTEST and PGTESTING env variable values are returning blank.
```shell
$ printenv |grep PG
PGTESTING=
PGTEST=
PGHOST=postgres
```

However, when I run `mise envs`, it shows the proper exports.
```
...
export PGHOST=postgres
export PGTEST=postgres
export PGTESTING=postgres
...
```

I realize this usage isn't covered in the docs, but thought I'd check and see if I need to configure things differently. For now, I'm just loading the env files with hooks, so I have a workaround.

Mise doctor output
```
mise doctor
version: 2024.12.10 linux-x64 (d52ec61 2024-12-14)
activated: yes
shims_on_path: no

build_info:
  Target: x86_64-unknown-linux-gnu
  Features: DEFAULT, NATIVE_TLS, OPENSSL
  Built: Sat, 14 Dec 2024 18:47:35 +0000
  Rust Version: rustc 1.83.0 (90b35a623 2024-11-26)
  Profile: release

shell:
  /bin/bash
  GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)
  Copyright (C) 2020 Free Software Foundation, Inc.
  License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

  This is free software; you are free to change and redistribute it.
  There is NO WARRANTY, to the extent permitted by law.

dirs:
  cache: ~/.cache/mise
  config: ~/.config/mise
  data: ~/.local/share/mise
  shims: ~/.local/share/mise/shims
  state: ~/.local/state/mise

config_files:
  ~/.config/mise/config.toml
  ~/src/example/.python-version
  ~/src/example/mise.toml

backends:
  aqua
  asdf
  cargo
  core
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:

toolset:
  aqua:BurntSushi/ripgrep@14.1.1
  aqua:astral-sh/uv@0.5.6
  aqua:caddyserver/caddy@2.8.4
  aqua:docker/compose@1.29.2
  core:node@22.12.0
  core:python@3.12.7

path:
  ~/.local/share/mise/installs/caddy/2.8.4
  ~/.local/share/mise/installs/docker-compose/1.29.2
  ~/.local/share/mise/installs/node/22.12.0/bin
  ~/.local/share/mise/installs/ripgrep/14.1.1/ripgrep-14.1.1-x86_64-unknown-linux-musl
  ~/.local/share/mise/installs/uv/0.5.6/uv-x86_64-unknown-linux-musl
  ~/.local/share/mise/installs/python/3.12.7/bin
  ~/.local/bin
  ~/bin
  /usr/local/sbin
  /usr/local/bin
  /usr/sbin
  /usr/bin
  /sbin
  /bin
  /usr/games
  /usr/local/games
  /snap/bin

env_vars:
  MISE_SHELL=bash

settings:
  experimental                          true                ~/src/example/mise.toml
  idiomatic_version_file_disable_tools  [".python_version"] ~/src/example/mise.toml

1 problem found:

1. new mise version 2024.12.24 available, currently on 2024.12.10
```

**Comment by jmoppel** [comment]
Update: I reread the docs again the [Multiple env._ Directives](https://mise.jdx.dev/environments/#multiple-env-directives) section, and tried using multiple `[[env]]` sections as shown below.

```toml
[[env]]
_.file = ".envs/.dev/.postgres"

[[env]]
_.file = ".envs/.dev/.local"
```

That still didn't seem to work either.

**Comment by syhol** [comment]
Something weird going on here. Sometimes I can recreate your problem and sometimes it works as you want.

I thought it might have something to do with the order the files are imported, either based on the updated/created timestamps or alphabetical, but I can't reliably reproduce anything.

**Comment by Marukome0743** [comment]
This has been fixed in **#11158** (merged), which will ship in the next release. 🎉

**What was happening**
Env vars loaded from `_.file` were inserted verbatim, so a value referencing a var defined in an *earlier* file (or an `[env]` block) collapsed to an empty string. dotenv files were hit hardest: `dotenvy` only substitutes `${VAR}` against the process env + same-file vars, so `${PGHOST}` from `.postgres` resolved to `""` in `.local`. That's also why `mise env` (which resolves everything together) showed the correct value while your actual shell was blank.

**What changes now** (with `env_shell_expand` enabled, the default):
- Values within a single `_.file` array now accumulate — a later file sees vars from an earlier one, matching how separate `[[env]]` blocks behave.
- Structured files (json/yaml/toml) keep `${VAR}` intact and go through the same `$VAR` engine used for `KEY = value` vars.
- dotenv files are parsed by `dotenvy` seeded with the accumulated env, so cross-file references like `${PGHOST}` resolve deterministically.

So your original example:
```
# .local
PGTEST="${PGHOST}"
PGTESTING=$PGHOST
```
will now correctly resolve `PGTEST` / `PGTESTING` to `postgres` in the shell, not just in `mise env`.

If you set `MISE_ENV_SHELL_EXPAND=false`, dotenv keeps its previous (unexpanded) behavior.

Thanks for the clear repro — it made this much easier to track down!

*This comment was generated with the help of an AI coding assistant.*
